sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/AppPreferencesRepositoryImpl.kt

```diff
@@ -0,0 +1,40 @@
+package co.nimblehq.sample.compose.data.repository
+
+import android.util.Log
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.*
+import co.nimblehq.sample.compose.domain.repository.AppPreferencesRepository
+import kotlinx.coroutines.flow.*
+import java.io.IOException
+import javax.inject.Inject
+
+class AppPreferencesRepositoryImpl @Inject constructor(
```

Comment:
@lydiasama Wait, I think we should not add `androidx.datastore:datastore-preferences` to the data module. Is it better to only add it to the `app` module and keep this AppPreferencesRepository implementation in the app module as well at `co.nimblehq.sample.compose.repository`?

This is the same when we need to integrate Firebase features such as FirebaseMessaging or FirebaseRemoteConfig. We need to put repository's implementations in the app module as well 💭

Reply:
@luongvo I considered the `Datastore` as a resource. So I think it makes sense to be in the data module. What do you think if we discuss about this in the discussion?

Reply:
If we consider in a standard Android project, it is common to place SharedPreferences in the app module, as it is often closely tied to the application's lifecycle and functions as a convenient storage location for user preferences and settings. This approach is straightforward and simple, as the SharedPreferences can be accessed directly from the Context of the application.

However, in a more modular or architectural approach, such as MVVM or Clean Architecture, it may be preferable to place the SharedPreferences in the data layer of the application. This approach allows for a clearer separation of concerns, as the data layer is responsible for managing and persisting data, including user preferences. Additionally, this makes it easier to test the data layer and provides a clear API for accessing and modifying SharedPreferences from the rest of the application. @luongvo

Reply:
@lydiasama Ok, so if we decide to put all data-related stuff in the `data` module, could we get rid of `androidx.datastore:datastore-preferences` in the app module?

Reply:
@luongvo We can't because for now, our project structure set the di inside of the `app` module. So we need to keep the dependency for it.

![Screenshot 2566-02-10 at 3 57 48 PM](https://user-images.githubusercontent.com/12026942/218048014-9d04e11b-0e5f-41a0-b2f8-84cf521e331e.png)

Reply:
@luongvo You can see the same structure with `RepositoryModule` that we call API. We implemented the `RepositoryModule` inside of `app` module, but `RepositoryImpl` is in the `data` module as well.

Reply:
@lydiasama No, it's not the same because the app module does not include retrofit, hence the app module should not include DataStore dependency too 💭

Reply:
@nimblehq/android-chapter What do you guys think?

Reply:
@luongvo I created a discussion [here](https://github.com/nimblehq/android-templates/discussions/406)

Reply:
@lydiasama Since we had a new discussion on this issue. I think we can handle this issue in part 2 or part 3 if needed https://github.com/nimblehq/android-templates/pull/374. 👌